File: packages/android/src/main/java/com/kozmos/components/POIDetailPanel/POIDetailPanel.kt
package com.kozmos.components.poidetailpanel

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bookmark
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.FavoriteBorder
import androidx.compose.material.icons.filled.Navigation
import androidx.compose.material.icons.filled.Place
import androidx.compose.material.icons.filled.Share
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.kozmos.components.button.KozmosButton
import com.kozmos.components.button.KozmosButtonVariant
import com.kozmos.components.iconbutton.KozmosIconButton
import com.kozmos.components.poimediagallery.KozmosPOIMediaGallery
import com.kozmos.contracts.KozmosPOIAccessRestrictions
import com.kozmos.contracts.KozmosPOIAction
import com.kozmos.contracts.KozmosPOIAvailability
import com.kozmos.contracts.KozmosPOIPresentation
import com.kozmos.contracts.KozmosPOIServicePresentation
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

/** Controlled state for a single POI action button. */
data class KozmosPOIActionState(
    val disabled: Boolean = false,
    val loading: Boolean = false,
    val pressed: Boolean = false,
    val message: String? = null,
    val messageTone: MessageTone = MessageTone.Status
) {
    enum class MessageTone {
        Status,
        Error
    }
}

enum class KozmosPOIDetailPanelPresentation {
    Inline,
    Sheet,
    Panel
}

/**
 * Full POI detail surface with media, actions, restrictions, and services.
 *
 * Mirrors the React `POIDetailPanel`. Only the actions listed by
 * [KozmosPOIPresentation.actions] are rendered, and every label is supplied
 * already localized.
 *
 * The web component's `titleLevel` prop has no counterpart here: Compose
 * semantics expose `heading()` as a boolean with no rank, so TalkBack cannot
 * distinguish an h2 from an h3. SwiftUI does support ranks and mirrors the prop
 * as `KozmosPOIDetailPanel.TitleLevel`.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun KozmosPOIDetailPanel(
    poi: KozmosPOIPresentation,
    actionLabels: Map<KozmosPOIAction, String>,
    onAction: (KozmosPOIAction, String) -> Unit,
    modifier: Modifier = Modifier,
    actionStates: Map<KozmosPOIAction, KozmosPOIActionState> = emptyMap(),
    onClose: (() -> Unit)? = null,
    closeLabel: String = "Close details",
    mediaLabel: String? = null,
    mediaPositionLabel: (Int, Int) -> String = { current, total -> "Image $current of $total" },
    accessRestrictionsHeading: String = "Access restrictions",
    servicesHeading: String = "Service options",
    presentation: KozmosPOIDetailPanelPresentation = KozmosPOIDetailPanelPresentation.Inline
) {
    val radius = KozmosDimensions.primitivesLayoutRadius300
    val shape = if (presentation == KozmosPOIDetailPanelPresentation.Sheet) {
        RoundedCornerShape(topStart = radius, topEnd = radius)
    } else {
        RoundedCornerShape(radius)
    }

    val showsAccessRestrictions = poi.accessRestrictions != null &&
        poi.accessRestrictions != KozmosPOIAccessRestrictions.None &&
        poi.accessRestrictionsLabel != null

    Surface(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = poi.name },
        shape = shape,
        color = KozmosColors.primitivesColorsBackground0,
        border = if (presentation == KozmosPOIDetailPanelPresentation.Sheet) {
            null
        } else {
            BorderStroke(1.dp, KozmosColors.primitivesColorsForeground300)
        },
        shadowElevation = if (presentation == KozmosPOIDetailPanelPresentation.Panel) 16.dp else 8.dp
    ) {
        // The body scrolls under a pinned header, which requires a bounded
        // height. When the caller nests the panel somewhere unbounded (another
        // scroll container, wrapContentSize) Compose would throw, so fall back
        // to growing naturally — the same way the web panel behaves.
        BoxWithConstraints(modifier = Modifier.fillMaxWidth()) {
        val bodyScrolls = constraints.hasBoundedHeight
        val bodyScrollState = rememberScrollState()

        Column(modifier = Modifier.fillMaxWidth()) {
            Header(poi = poi, onClose = onClose, closeLabel = closeLabel)

            Divider(color = KozmosColors.primitivesColorsForeground300)

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .then(
                        if (bodyScrolls) {
                            Modifier
                                .weight(1f, fill = false)
                                .verticalScroll(bodyScrollState)
                        } else {
                            Modifier
                        }
                    )
                    .padding(KozmosDimensions.primitivesLayoutSpacing200),
                verticalArrangement = Arrangement.spacedBy(
                    KozmosDimensions.primitivesLayoutSpacing200
                )
            ) {
                poi.description?.let { description ->
                    Text(
                        text = description,
                        style = MaterialTheme.typography.bodyMedium,
                        color = KozmosColors.primitivesColorsForeground500
                    )
                }

                if (poi.actions.isNotEmpty()) {
                    FlowRow(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(
                            KozmosDimensions.primitivesLayoutSpacing100
                        ),
                        verticalArrangement = Arrangement.spacedBy(
                            KozmosDimensions.primitivesLayoutSpacing100
                        )
                    ) {
                        poi.actions.forEach { action ->
                            val state = actionStates[action]
                            val isToggle = action == KozmosPOIAction.Favourite ||
                                action == KozmosPOIAction.Bookmark

                            KozmosButton(
                                onClick = { onAction(action, poi.id) },
                                modifier = Modifier.semantics {
                                    if (isToggle) selected = state?.pressed ?: false
                                },
                                variant = if (action == KozmosPOIAction.Navigate) {
                                    KozmosButtonVariant.Default
                                } else {
                                    KozmosButtonVariant.Outline
                                },
                                enabled = !(state?.disabled ?: false),
                                isLoading = state?.loading ?: false
                            ) {
                                Icon(
                                    imageVector = actionIcon(action),
                                    contentDescription = null,
                                    modifier = Modifier.size(16.dp)
                                )
                                Text(actionLabels[action] ?: action.value)
                            }
                        }
                    }
                }

                poi.actions.forEach { action ->
                    val state = actionStates[action]
                    val message = state?.message
                    if (message != null) {
                        Text(
                            text = message,
                            style = MaterialTheme.typography.bodyMedium,
                            color = if (state.messageTone == KozmosPOIActionState.MessageTone.Error) {
                                KozmosColors.primitivesColorsEmotionalDanger600
                            } else {
                                KozmosColors.primitivesColorsForeground100
                            },
                            modifier = Modifier
                                .fillMaxWidth()
                                .clip(RoundedCornerShape(KozmosDimensions.primitivesLayoutRadius100))
                                .background(KozmosColors.primitivesColorsBackground100)
                                .padding(
                                    horizontal = KozmosDimensions.primitivesLayoutSpacing150,
                                    vertical = KozmosDimensions.primitivesLayoutSpacing100
                                )
                                .semantics { liveRegion = LiveRegionMode.Polite }
                        )
                    }
                }

                if (showsAccessRestrictions) {
                    val accessRestrictionsLabel = poi.accessRestrictionsLabel.orEmpty()
                    Surface(
                        modifier = Modifier
                            .fillMaxWidth()
                            .semantics {
                                contentDescription =
                                    "$accessRestrictionsHeading, $accessRestrictionsLabel"
                            },
                        shape = RoundedCornerShape(KozmosDimensions.primitivesLayoutRadius100),
                        color = KozmosColors.primitivesColorsBackground100.copy(alpha = 0.4f),
                        border = BorderStroke(1.dp, KozmosColors.primitivesColorsForeground300)
                    ) {
                        Text(
                            text = accessRestrictionsLabel,
                            style = MaterialTheme.typography.bodyMedium,
                            color = KozmosColors.primitivesColorsForeground100,
                            modifier = Modifier.padding(KozmosDimensions.primitivesLayoutSpacing150)
                        )
                    }
                }

                KozmosPOIMediaGallery(
                    media = poi.media,
                    label = mediaLabel ?: "${poi.name} photos",
                    positionLabel = mediaPositionLabel
                )

                val services = poi.services
                if (!services.isNullOrEmpty()) {
                    Services(heading = servicesHeading, services = services)
                }
            }
        }
        }
    }
}

private fun actionIcon(action: KozmosPOIAction): ImageVector = when (action) {
    KozmosPOIAction.Navigate -> Icons.Default.Navigation
    KozmosPOIAction.Favourite -> Icons.Default.FavoriteBorder
    KozmosPOIAction.Bookmark -> Icons.Default.Bookmark
    KozmosPOIAction.Share -> Icons.Default.Share
    KozmosPOIAction.Order -> Icons.Default.ShoppingCart
}

@Composable
private fun Header(
    poi: KozmosPOIPresentation,
    onClose: (() -> Unit)?,
    closeLabel: String
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(KozmosDimensions.primitivesLayoutSpacing200),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing150)
    ) {
        POILogo(poi = poi)

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing50)
        ) {
            Text(
                text = poi.name,
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold,
                color = KozmosColors.primitivesColorsForeground100,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.semantics { heading() }
            )

            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(
                    KozmosDimensions.primitivesLayoutSpacing50
                )
            ) {
                Icon(
                    imageVector = Icons.Default.Place,
                    contentDescription = null,
                    tint = KozmosColors.primitivesColorsForeground500,
                    modifier = Modifier.size(16.dp)
                )
                Text(
                    text = poi.locationLabel,
                    style = MaterialTheme.typography.bodyMedium,
                    color = KozmosColors.primitivesColorsForeground500,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }

            poi.availabilityLabel?.let { availabilityLabel ->
                Text(
                    text = availabilityLabel,
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = if (poi.availability == KozmosPOIAvailability.Open) {
                        KozmosColors.componentsPrimaryButtonsSuccessButtonBackgroundIdle
                    } else {
                        KozmosColors.primitivesColorsForeground500
                    },
                    modifier = Modifier.semantics {
                        contentDescription = "Availability: $availabilityLabel"
                    }
                )
            }
        }

        if (onClose != null) {
            KozmosIconButton(
                icon = Icons.Default.Close,
                onClick = onClose,
                contentDescription = closeLabel
            )
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun Services(
    heading: String,
    services: List<KozmosPOIServicePresentation>
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .semantics { contentDescription = heading },
        verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100)
    ) {
        Text(
            text = heading,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.SemiBold,
            color = KozmosColors.primitivesColorsForeground100,
            modifier = Modifier.semantics { heading() }
        )

        FlowRow(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(
                KozmosDimensions.primitivesLayoutSpacing100
            ),
            verticalArrangement = Arrangement.spacedBy(
                KozmosDimensions.primitivesLayoutSpacing100
            )
        ) {
            services.forEach { service ->
                Surface(
                    shape = RoundedCornerShape(percent = 50),
                    color = KozmosColors.primitivesColorsBackground0,
                    border = BorderStroke(1.dp, KozmosColors.primitivesColorsForeground300)
                ) {
                    Text(
                        text = service.label,
                        style = MaterialTheme.typography.bodyMedium,
                        color = KozmosColors.primitivesColorsForeground100,
                        modifier = Modifier.padding(
                            horizontal = KozmosDimensions.primitivesLayoutSpacing150,
                            vertical = KozmosDimensions.primitivesLayoutSpacing100
                        )
                    )
                }
            }
        }
    }
}

@Composable
private fun POILogo(poi: KozmosPOIPresentation) {
    val logo = poi.logo
    val shape = RoundedCornerShape(KozmosDimensions.primitivesLayoutRadius100)

    if (logo != null) {
        AsyncImage(
            model = logo.src,
            contentDescription = logo.alt,
            contentScale = ContentScale.Fit,
            modifier = Modifier
                .size(48.dp)
                .clip(shape)
        )
    } else {
        Box(
            modifier = Modifier
                .size(48.dp)
                .clip(shape)
                .background(KozmosColors.primitivesColorsBackground100),
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = poi.logoFallbackInitial,
                style = MaterialTheme.typography.bodyLarge,
                fontWeight = FontWeight.Bold,
                color = KozmosColors.primitivesColorsForeground500
            )
        }
    }
}

File: packages/ios/Sources/Components/POIDetailPanel/POIDetailPanel.swift
import SwiftUI

/// Controlled state for a single POI action button.
public struct KozmosPOIActionState: Sendable, Hashable {
    public enum MessageTone: Sendable, Hashable {
        case status
        case error
    }

    public let disabled: Bool
    public let loading: Bool
    public let pressed: Bool
    public let message: String?
    public let messageTone: MessageTone

    public init(
        disabled: Bool = false,
        loading: Bool = false,
        pressed: Bool = false,
        message: String? = nil,
        messageTone: MessageTone = .status
    ) {
        self.disabled = disabled
        self.loading = loading
        self.pressed = pressed
        self.message = message
        self.messageTone = messageTone
    }
}

/// Native POI card. Presentation is controlled by the product; no SDK objects or
/// taxonomy lookups are embedded in this view. Existing basic callers still work.
public struct KozmosPOIDetailPanel: View {
    public enum Presentation { case inline, sheet, panel }
    public enum TitleLevel {
        case h2, h3
        var accessibilityHeadingLevel: AccessibilityHeadingLevel { self == .h2 ? .h2 : .h3 }
    }

    private let poi: KozmosPOIPresentation
    private let details: KozmosPOIDetailsPresentation
    private let actionLabels: [KozmosPOIAction: String]
    private let actionStates: [KozmosPOIAction: KozmosPOIActionState]
    private let supplementaryActionStates: [String: KozmosPOIActionState]
    private let closeLabel: String
    private let mediaLabel: String
    private let mediaPositionLabel: (Int, Int) -> String
    private let mediaPreviousLabel: String
    private let mediaNextLabel: String
    private let mediaUnavailableLabel: String
    private let mediaControlsLabel: String?
    private let mediaRetryHint: String
    private let accessRestrictionsHeading: String
    private let servicesHeading: String
    private let readMoreLabel: String
    private let readLessLabel: String
    private let tagsLabel: String
    private let loadingLabel: String
    private let presentation: Presentation
    private let titleLevel: TitleLevel
    private let onAction: (KozmosPOIAction, String) -> Void
    private let onSupplementaryAction: ((String, String) -> Void)?
    private let onClose: (() -> Void)?

    public init(
        poi: KozmosPOIPresentation,
        actionLabels: [KozmosPOIAction: String],
        onAction: @escaping (KozmosPOIAction, String) -> Void,
        actionStates: [KozmosPOIAction: KozmosPOIActionState] = [:],
        onClose: (() -> Void)? = nil,
        closeLabel: String = "Close details",
        mediaLabel: String? = nil,
        mediaPositionLabel: @escaping (Int, Int) -> String = { "Image \($0) of \($1)" },
        mediaPreviousLabel: String = "Previous image",
        mediaNextLabel: String = "Next image",
        mediaUnavailableLabel: String = "Image unavailable",
        mediaControlsLabel: String? = nil,
        mediaRetryHint: String = "Double-tap to try again",
        accessRestrictionsHeading: String = "Access restrictions",
        servicesHeading: String = "Service options",
        presentation: Presentation = .inline,
        titleLevel: TitleLevel = .h2,
        details: KozmosPOIDetailsPresentation = .init(),
        supplementaryActionStates: [String: KozmosPOIActionState] = [:],
        onSupplementaryAction: ((String, String) -> Void)? = nil,
        readMoreLabel: String = "Read more",
        readLessLabel: String = "Read less",
        tagsLabel: String = "Tags",
        loadingLabel: String = "Loading"
    ) {
        self.poi = poi; self.details = details; self.actionLabels = actionLabels
        self.onAction = onAction; self.actionStates = actionStates; self.onClose = onClose
        self.closeLabel = closeLabel; self.mediaLabel = mediaLabel ?? "\(poi.name) photos"
        self.mediaPositionLabel = mediaPositionLabel
        self.mediaPreviousLabel = mediaPreviousLabel; self.mediaNextLabel = mediaNextLabel
        self.mediaUnavailableLabel = mediaUnavailableLabel; self.mediaControlsLabel = mediaControlsLabel
        self.mediaRetryHint = mediaRetryHint
        self.accessRestrictionsHeading = accessRestrictionsHeading; self.servicesHeading = servicesHeading
        self.presentation = presentation; self.titleLevel = titleLevel
        self.supplementaryActionStates = supplementaryActionStates
        self.onSupplementaryAction = onSupplementaryAction
        self.readMoreLabel = readMoreLabel; self.readLessLabel = readLessLabel; self.tagsLabel = tagsLabel
        self.loadingLabel = loadingLabel
    }

    private var quickActions: [KozmosPOIAction] {
        poi.actions.filter { $0 == .favourite || $0 == .bookmark }
    }
    private var stripActions: [KozmosPOIAction] {
        poi.actions.filter { $0 != .favourite && $0 != .bookmark }
    }

    private static func systemImage(for action: KozmosPOIAction) -> String {
        switch action {
        case .navigate: return "location"
        case .favourite: return "heart"
        case .bookmark: return "bookmark"
        case .share: return "square.and.arrow.up"
        case .order: return "bag"
        }
    }

    public var body: some View {
        // Yields to the shell's sheet: scrolls only at its largest detent, and
        // hands a downward drag back to the sheet once at its top.
        KozmosPanelScrollView {
            VStack(alignment: .leading, spacing: 0) {
                header.padding(16)
                if let description = poi.description, !description.isEmpty {
                    Text(description)
                        .font(KozmosTypography.subheadline)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.horizontal, 16).padding(.bottom, 16)
                }
                if !stripActions.isEmpty || !details.supplementaryActions.isEmpty {
                    // The sheet's smallest detent rests on this row: the
                    // prototype's place card peeks at its header and Go.
                    actionButtons.padding(.bottom, 16).kozmosPanelPeekAnchor()
                }
                messages
                if !details.visibleSummary.isEmpty {
                    POIDetailSummary(items: details.visibleSummary)
                }
                VStack(alignment: .leading, spacing: 20) {
                    if let restriction = poi.accessRestrictions,
                       restriction != KozmosPOIAccessRestrictions.none,
                       let label = poi.accessRestrictionsLabel {
                        Text(label).font(KozmosTypography.subheadline)
                            .padding(12).frame(maxWidth: .infinity, alignment: .leading)
                            .background(insetSurface)
                            .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl))
                            .accessibilityLabel("\(accessRestrictionsHeading), \(label)")
                    }
                    KozmosPOIMediaGallery(
                        surface: insetSurface,
                        media: poi.media, label: mediaLabel, positionLabel: mediaPositionLabel,
                        previousLabel: mediaPreviousLabel, nextLabel: mediaNextLabel,
                        controlsLabel: mediaControlsLabel, unavailableLabel: mediaUnavailableLabel,
                        loadingLabel: loadingLabel, retryHint: mediaRetryHint)
                    if let services = poi.services, !services.isEmpty {
                        VStack(alignment: .leading, spacing: 8) {
                            Text(servicesHeading).font(KozmosTypography.footnote)
                                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                                .accessibilityAddTraits(.isHeader)
                            POIDetailTags(items: services.map(KozmosPOIDetailTag.init(service:)))
                        }
                    }
                    POIDetailExtendedContent(details: details, readMoreLabel: readMoreLabel,
                                             readLessLabel: readLessLabel, tagsLabel: tagsLabel)
                }
                .padding(16)
            }
        }
        // Reset scroll/disclosure state only when selecting a different place.
        .id(poi.id)
        .foregroundColor(KozmosColors.primitivesColorsForeground100)
        // In a sheet the panel paints no surface of its own: it sits on the
        // sheet's, as the browse panel does, with no border and no card.
        .background(presentation == .sheet ? Color.clear : KozmosColors.primitivesColorsBackground0)
        .clipShape(panelShape)
        .overlay(panelShape.stroke(presentation == .sheet ? Color.clear : KozmosColors.semanticsBorderSubtle, lineWidth: 1))
        .accessibilityElement(children: .contain)
        .accessibilityLabel(poi.name)
    }

    /// An inset block's surface: the muted grey on the panel's own white, and
    /// white on a sheet, whose surface is that grey.
    private var insetSurface: Color {
        presentation == .sheet ? KozmosColors.primitivesColorsBackground0 : KozmosColors.primitivesColorsBackground100
    }

    private var panelShape: KozmosPanelShape {
        KozmosPanelShape(radius: KozmosDimensions.semanticsRadiusControl, roundsBottom: presentation != .sheet)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 12) {
            // The name and the quick buttons share one row whatever the name's
            // length: a long name wraps beside them, three lines at most, and
            // never pushes them under it.
            HStack(alignment: .top, spacing: 8) { identity; quickButtons }
            ViewThatFits(in: .horizontal) {
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    location; Spacer(minLength: 4); availability
                }
                VStack(alignment: .leading, spacing: 4) { location; availability }
            }
        }
    }

    private var identity: some View {
        HStack(alignment: .top, spacing: 8) {
            if let logo = poi.logo {
                AsyncImage(url: POIDetailIcon.remoteURL(logo.src)) { phase in
                    if let image = phase.image {
                        image.resizable().scaledToFit()
                    } else {
                        Text(poi.logoFallbackInitial).font(KozmosTypography.headline)
                            .frame(maxWidth: .infinity, maxHeight: .infinity)
                            .background(insetSurface)
                    }
                }
                .frame(width: 48, height: 48)
                .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl))
                .accessibilityLabel(logo.alt)
            }
            Text(poi.name).font(KozmosTypography.title3.weight(.semibold))
                .lineLimit(3)
                .fixedSize(horizontal: false, vertical: true)
                .accessibilityAddTraits(.isHeader)
                .accessibilityHeading(titleLevel.accessibilityHeadingLevel)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var location: some View {
        Text(poi.locationLabel).font(KozmosTypography.subheadline)
            .foregroundColor(KozmosColors.primitivesColorsForeground500)
            .fixedSize(horizontal: false, vertical: true)
    }

    @ViewBuilder private var availability: some View {
        if let label = poi.availabilityLabel {
            Text(label).font(KozmosTypography.caption.weight(.semibold))
                .foregroundColor(poi.availability == .open
                    ? KozmosColors.componentsPrimaryButtonsSuccessButtonBackgroundIdle
                    : KozmosColors.primitivesColorsForeground500)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    private var quickButtons: some View {
        HStack(spacing: 6) {
            ForEach(quickActions, id: \.self) { action in
                POIDetailActionButton(label: actionLabels[action] ?? action.rawValue,
                                     systemImage: Self.systemImage(for: action), iconOnly: true,
                                     loadingLabel: loadingLabel,
                                     state: actionStates[action] ?? .init()) { onAction(action, poi.id) }
                    .accessibilityIdentifier("poi-action-\(action.rawValue)")
            }
            if let onClose {
                POIDetailActionButton(label: closeLabel, systemImage: "xmark", iconOnly: true, action: onClose)
                    .accessibilityIdentifier("poi-close")
            }
        }
    }

    private var actionButtons: some View {
        ScrollView(.horizontal) {
            HStack(spacing: 8) {
                ForEach(stripActions, id: \.self) { action in
                    POIDetailActionButton(
                        label: actionLabels[action] ?? action.rawValue,
                        systemImage: Self.systemImage(for: action),
                        estimate: action == .navigate ? details.travelEstimate.map {
                            [$0.durationLabel, $0.distanceLabel].compactMap { $0 }.joined(separator: " · ")
                        } : nil,
                        primary: action == .navigate,
                        loadingLabel: loadingLabel,
                        state: actionStates[action] ?? .init()
                    ) { onAction(action, poi.id) }
                    .accessibilityIdentifier("poi-action-\(action.rawValue)")
                }
                ForEach(details.supplementaryActions) { item in
                    POIDetailActionButton(label: item.label, systemImage: item.systemImage,
                                         loadingLabel: loadingLabel,
                                         state: supplementaryActionStates[item.action] ?? .init(disabled: onSupplementaryAction == nil)) {
                        onSupplementaryAction?(item.action, poi.id)
                    }
                    .disabled(onSupplementaryAction == nil)
                    .accessibilityIdentifier("poi-action-\(item.action)")
                }
            }
            .padding(.horizontal, 16).padding(.vertical, 2)
        }
        .accessibilityIdentifier("poi-actions")
    }

    private var allMessages: [(String, KozmosPOIActionState)] {
        poi.actions.compactMap { action in actionStates[action].map { ("core:\(action.rawValue)", $0) } }
        + details.supplementaryActions.compactMap { item in supplementaryActionStates[item.action].map { ("supplementary:\(item.action)", $0) } }
    }

    private var messages: some View {
        ForEach(allMessages.filter { $0.1.message != nil }, id: \.0) { _, state in
            if let message = state.message {
                Text(message).font(KozmosTypography.subheadline)
                    .foregroundColor(state.messageTone == .error
                        ? KozmosColors.primitivesColorsEmotionalDanger600
                        : KozmosColors.primitivesColorsForeground100)
                    .padding(12).frame(maxWidth: .infinity, alignment: .leading)
                    .background(insetSurface)
                    .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl))
                    .padding(.horizontal, 16).padding(.bottom, 12)
                    .accessibilityAddTraits(.updatesFrequently)
            }
        }
    }
}

struct POIDetailActionButton: View {
    let label: String
    let systemImage: String?
    var estimate: String? = nil
    var primary = false
    var iconOnly = false
    var loadingLabel = "Loading"
    var state = KozmosPOIActionState()
    let action: () -> Void

    private var filled: Bool { primary || state.pressed }
    private var foreground: Color {
        filled ? KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
               : KozmosColors.primitivesColorsForeground100
    }

    var body: some View {
        Button(action: action) {
            HStack(spacing: 8) {
                if state.loading {
                    ProgressView().controlSize(.small).tint(foreground)
                } else if let systemImage {
                    Image(systemName: systemImage)
                        .font(primary ? .title3 : .body)
                        .accessibilityHidden(true)
                }
                if !iconOnly {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(label).font(primary ? KozmosTypography.font(.callout).weight(.semibold) : KozmosTypography.subheadline)
                        if let estimate { Text(estimate).font(KozmosTypography.caption2) }
                    }
                }
            }
            .foregroundColor(foreground)
            .padding(.horizontal, iconOnly ? 0 : 16).padding(.vertical, 8)
            .frame(minWidth: 44, minHeight: primary && estimate != nil ? 56 : 44)
            .background(filled ? KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle : Color.clear)
            .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl))
            .overlay(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl)
                .stroke(filled ? Color.clear : KozmosColors.semanticsBorderSubtle, lineWidth: 1))
        }
        .buttonStyle(.plain)
        .disabled(state.disabled || state.loading)
        .opacity(state.disabled ? 0.5 : 1)
        .accessibilityLabel(label)
        .accessibilityValue(state.loading ? Text(loadingLabel) : Text(estimate ?? ""))
        .accessibilityAddTraits(state.pressed ? [.isSelected] : [])
    }
}

/// Panel outline that can drop its bottom corners for sheet presentation.
///
/// `UnevenRoundedRectangle` would express this directly but is iOS 17+, and the
/// package deploys to iOS 16.
struct KozmosPanelShape: InsettableShape {
    var radius: CGFloat
    var roundsBottom: Bool
    var insetAmount: CGFloat = 0

    func inset(by amount: CGFloat) -> KozmosPanelShape {
        var shape = self
        shape.insetAmount += amount
        return shape
    }

    func path(in outer: CGRect) -> Path {
        let rect = outer.insetBy(dx: insetAmount, dy: insetAmount)
        let r = min(radius, min(rect.width, rect.height) / 2)
        let bottomR = roundsBottom ? r : 0

        var path = Path()
        path.move(to: CGPoint(x: rect.minX, y: rect.minY + r))
        path.addArc(
            center: CGPoint(x: rect.minX + r, y: rect.minY + r),
            radius: r,
            startAngle: .degrees(180),
            endAngle: .degrees(270),
            clockwise: false
        )
        path.addLine(to: CGPoint(x: rect.maxX - r, y: rect.minY))
        path.addArc(
            center: CGPoint(x: rect.maxX - r, y: rect.minY + r),
            radius: r,
            startAngle: .degrees(270),
            endAngle: .degrees(0),
            clockwise: false
        )
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY - bottomR))

        if bottomR > 0 {
            path.addArc(
                center: CGPoint(x: rect.maxX - bottomR, y: rect.maxY - bottomR),
                radius: bottomR,
                startAngle: .degrees(0),
                endAngle: .degrees(90),
                clockwise: false
            )
        }

        path.addLine(to: CGPoint(x: rect.minX + bottomR, y: rect.maxY))

        if bottomR > 0 {
            path.addArc(
                center: CGPoint(x: rect.minX + bottomR, y: rect.maxY - bottomR),
                radius: bottomR,
                startAngle: .degrees(90),
                endAngle: .degrees(180),
                clockwise: false
            )
        }

        path.closeSubpath()
        return path
    }
}

/// Wrapping layout for property/service chips (actions scroll horizontally).
///
/// SwiftUI has no built-in flow container, and these rows must wrap rather than
/// clip when localized labels or Dynamic Type make them wide.
struct FlowLayout: Layout {
    var spacing: CGFloat

    init(spacing: CGFloat = 8) {
        self.spacing = spacing
    }

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        let rows = layoutRows(maxWidth: maxWidth, subviews: subviews)
        let height = rows.reduce(into: CGFloat.zero) { total, row in
            total += row.height
        } + spacing * CGFloat(max(rows.count - 1, 0))
        let width = rows.map(\.width).max() ?? 0
        return CGSize(width: min(width, maxWidth), height: height)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let rows = layoutRows(maxWidth: bounds.width, subviews: subviews)
        var y = bounds.minY

        for row in rows {
            var x = bounds.minX
            for index in row.indices {
                let size = subviews[index].sizeThatFits(.init(width: bounds.width, height: nil))
                subviews[index].place(
                    at: CGPoint(x: x, y: y),
                    proposal: ProposedViewSize(size)
                )
                x += size.width + spacing
            }
            y += row.height + spacing
        }
    }

    private struct Row {
        var indices: [Int] = []
        var width: CGFloat = 0
        var height: CGFloat = 0
    }

    private func layoutRows(maxWidth: CGFloat, subviews: Subviews) -> [Row] {
        var rows: [Row] = []
        var current = Row()

        for index in subviews.indices {
            let size = subviews[index].sizeThatFits(.init(width: maxWidth.isFinite ? maxWidth : nil, height: nil))
            let projected = current.indices.isEmpty ? size.width : current.width + spacing + size.width

            if projected > maxWidth, !current.indices.isEmpty {
                rows.append(current)
                current = Row()
                current.indices = [index]
                current.width = size.width
                current.height = size.height
            } else {
                current.indices.append(index)
                current.width = projected
                current.height = max(current.height, size.height)
            }
        }

        if !current.indices.isEmpty { rows.append(current) }
        return rows
    }
}
